### PR TITLE
Move immutable.re component to bs-containers

### PR DESCRIPTION
**Pull request by dorafmon**
Move some core interface from `Immutable.re` to `bs-containers`

### CODE REVIEW

**Comment by dorafmon** [comment]
@bordoley, could you also review this please?

**Comment by jordwalke** [comment]
Is the thinking here to be standardizing on the common parts even if a formal dependency isn't formed between the two libraries initially?

File: src/bopt.mli [hunk 1]
```diff
@@ -47,7 +47,7 @@ val equal : ('a -> 'a -> bool) -> 'a t -> 'a t -> bool
 (** [equal p a b] is [true] if both are [None], [p x y] if both are [Some _],
     [false] otherwise *)
 
-val compare : ('a -> 'a -> Comparison.comparison) -> 'a t -> 'a t -> Comparison.comparison
+val compare : ('a -> 'a -> TotalOrder.t) -> 'a t -> 'a t -> TotalOrder.t
```
Comment (bordoley) [inline src/bopt.mli]: Why not:
```
include Comparable.S with type t 'a := option 'a;
```

File: src/comparator.ml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+type 'a t = 'a -> 'a -> TotalOrder.t
+
+let make (compare : 'a -> 'a -> TotalOrder.t) (this : 'a) (that : 'a): TotalOrder.t =
```
Comment (bordoley) [inline src/comparator.ml]: I don't quite grok the purpose of this function?

File: src/totalOrder.ml [hunk 3]
```diff
@@ -1,14 +1,10 @@
-type comparison =
+(** A type represent a total order relationship *)
```
Comment (bordoley) [inline src/totalOrder.ml]: Why totalOrder vs. ordering? Not opposed, just curious to the rationale. For reference, I took the name from Haskell: https://hackage.haskell.org/package/base-4.9.1.0/docs/Data-Ord.html

File: src/totalOrder.ml [hunk 4]
```diff
@@ -1,14 +1,10 @@
...
+
+type t =
   | Greater
   | Less
```
Comment (bordoley) [inline src/totalOrder.ml]: I'm ok exposing the actual variant in the public type definition, but it's also important to expose values as Imm.re does to enable usage from js.

**Comment by dorafmon** [comment]
@jordwalke Hi, so what I am trying to do is  1) step one to move some core concept/interface to this repo and make it a self contained core library for both `immutable.re` and later `bs-containers`; 2) after step one we can think about how to make this repo a dependency for `immutable.re`,  be it using npm/opam/yarn.  So yes it is not formal dependency for `immutable.re` yet but it will be later.

cc. @bordoley is this approach okay with you?

File: src/comparator.ml (hunk 2 above)
Comment (dorafmon) [inline src/comparator.ml]: I think it should be 
``
let make (compare: 'a -> 'a -> TotalOrder.t) : 'a t = compare
``

I will update the PR later.

Also re. @glennsl .

File: src/totalOrder.ml (hunk 3 above)
Comment (dorafmon) [inline src/totalOrder.ml]: I think a sensible name would be `Ord`, will change that.

File: src/totalOrder.ml (hunk 4 above)
Comment (dorafmon) [inline src/totalOrder.ml]: Ah, now I understand why you expose those functions.

File: src/totalOrder.ml (hunk 3 above)
Comment (bordoley) [inline src/totalOrder.ml]: In general, in imm.re i've avoid abbreviated names, hence I've used Ordering, Sequence, etc. In general i feel this optimizes for readability over terseness.

**Comment by bordoley** [comment]
I agree with the general approach of letting imm.re (now officially 1 week old) and bs-base/containers develop in parallel, keeping each other mind. It's too early to do a full project merge but hopefully we get base to a point were it forms a solid xplat lib and imm.re can migrate to using it as a base type lib.

File: src/comparator.ml (hunk 2 above)
Comment (dorafmon) [inline src/comparator.ml]: I took it from [here](https://github.com/facebookincubator/immutable-re/blob/master/immutable/src/core/Comparator.re#L12). I guess the intention is provide a way to wrap a simple comparison function into a `Comparator.t`, but I can remove it if you think it is too trivial.

Comment (bordoley) [inline src/comparator.ml]: The intent of that 'make' function is to take a traditional ocaml compare function which returns an int and convert it into an Immutable.re compatible Comparator function which returns an Ordering. 
```
type t 'a = 'a => 'a => Ordering.t;

let make (compare: 'a => 'a => int) (this: 'a) (that: 'a): Ordering.t => {
  let cmp = compare this that;

  if (cmp > 0) Ordering.greaterThan
  else if (cmp < 0) Ordering.lessThan
  else Ordering.equal;
};
```

Comment (dorafmon) [inline src/comparator.ml]: Then it can be replaced by the `fromInt` function from the `Ordering` module.

Comment (bordoley) [inline src/comparator.ml]: fromInt is less convenient. make is intended to return a (Comparator.t 'a) although I don't think I even expose it in the public api.

**Comment by dorafmon** [comment]
@bordoley I updated the PR, hopefully, this interface is a bit more clear. Could you review please? Thanks!

File: src/totalOrder.ml (hunk 3 above)
Comment (glennsl) [inline src/totalOrder.ml]: +1 :)

File: src/comparator.ml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+type 'a t = 'a -> 'a -> Ordering.t
+
+let make (compare: 'a -> 'a -> int) : 'a t = fun this that ->
+  let r = compare this that in if r = 0 then Ordering.Equal
+  else begin
+    if r > 0 then Ordering.Greater
+    else Ordering.Less
+  end 
+
+let makeDefault (this: 'a) (that: 'a): Ordering.t =
```
Comment (glennsl) [inline src/comparator.ml]: I still don't understand this one. It doesn't really make anything, even if partially applied. Could it just be `default`?.

File: src/comparator.ml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+type 'a t = 'a -> 'a -> Ordering.t
+
+let make (compare: 'a -> 'a -> int) : 'a t = fun this that ->
+  let r = compare this that in if r = 0 then Ordering.Equal
```
Comment (glennsl) [inline src/comparator.ml]: Nitpick, the ìf` should really be at the same level as the branches

File: src/comparable.ml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+module type S = sig
+  type t
+
+  include Equatable.S with type t := t
+
+  val compare:t Comparator.t
+end
+
+module Default(T: sig type t end) : S = struct
```
Comment (dorafmon) [inline src/comparable.ml]: @glennsl does this make it more clearer?

Comment (bordoley) [inline src/comparable.ml]: Can I highly recommend not including structural equality and comparison in bs-base. In practice it's very slow, and almost always the wrong choice.

Comment (dorafmon) [inline src/comparable.ml]: Sorry for my limited knowledge here. So what is the right alternative? Thanks.

Comment (bordoley) [inline src/comparable.ml]: Provide a standard comparators and equality functions for all the built in types (int, string, etc). Authors should roll their own comparison and equality functions for their types. Polymorphic compare requires walking the object tree at runtime.  See: https://janestreet.github.io/ocaml-perf-notes.html

Ideally we could use the type information to generate comparison functions per type at compile time (maybe using a PPX transform) that uses the type info to generate a static comparator for record and union types. This is exactly how F# works.

Comment (dorafmon) [inline src/comparable.ml]: yeah I was thinking integrating `deriving`.

Comment (bordoley) [inline src/comparable.ml]: Another data point, in order to do a fair performance comparison of Immutable.re's IntSet vs. an ocaml Set of ints, I had to do the following:

https://github.com/facebookincubator/immutable-re/blob/master/immutable/perf/SetPerfTester.re#L60-L73

Using polymorphic compare in CamlIntSet was giving IntSet a significant unfair advantage.

Comment (dorafmon) [inline src/comparable.ml]: Well, it is late in London now. Let me update this tomorrow, I have some idea we can try.

**Comment by dorafmon** [comment]
@glennsl @bordoley I think this is ready to merge, could you comment? Thanks. I moved some core concepts from Immutable.re to `src/core` as how they are now.

**Comment by bordoley** [comment]
Looks good to me. Some unrelated changes in the PR, not sure if that was intentional or not.

File: src/barray.mli [hunk 8]
```diff
@@ -11,131 +11,131 @@ external length : 'a array -> int = "%array_length"
...
 (** [Array.get a n] returns the element number [n] of array [a].
-   The first element has number 0.
-   The last element has number [Array.length a - 1].
-   You can also write [a.(n)] instead of [Array.get a n].
-   Raise [Invalid_argument "index out of bounds"]
-   if [n] is outside the range 0 to [(Array.length a - 1)]. *)
+    The first element has number 0.
+    The last element has number [Array.length a - 1].
+    You can also write [a.(n)] instead of [Array.get a n].
+    Raise [Invalid_argument "index out of bounds"]
```
Comment (glennsl) [inline src/barray.mli]: Just a note for the future: These should use the `@raise` doctag

**Comment by glennsl** [comment]
Looks good to me too. Merging...

**Pull request merged**